src/binary_basic.ml

```diff
@@ -0,0 +1,471 @@
+open Types
+
+module Input = struct
+  type t =
+    { bytes : bytes
+    ; pt : int
+    ; size : int
+    }
+
+  let _pp input =
+    print_int input.pt;
+    print_string " ";
+    print_int input.size;
+    print_newline ()
+
+  let is_empty input = input.size = 0
+
+  let from_bytes bytes =
+    let size = Bytes.length bytes in
+    { bytes; pt = 0; size }
+
+  let sub ~pos ~len input =
+    if pos <= input.size && len <= input.size - pos then
+      { input with pt = input.pt + pos; size = len }
+    else failwith "Input.sub error"
+
+  let sub_suffix pos input = sub ~pos ~len:(input.size - pos) input
+
+  let sub_prefix len input = sub ~pos:0 ~len input
+
+  let get n input =
+    if n < input.size then Some (Bytes.get input.bytes (input.pt + n)) else None
+
+  let get0 = get 0
+end
+
+let read_byte input =
+  match Input.get0 input with
+  | None -> failwith "read_byte error"
+  | Some c ->
+    let next_input = Input.sub_suffix 1 input in
+    (c, next_input)
+
+(* TODO: U32 spec implem. For now, only 1 byte *)
+let read_U32 input =
+  match Input.get0 input with
+  | None -> failwith "read_U32 error"
+  | Some c ->
+    let next_input = Input.sub_suffix 1 input in
+    let i = Char.code c in
+    let i = if i < 128 then i else 0 (* 0 : todo *) in
+    (* 128=2⁷ *)
+    (i, next_input)
+
+let read_U64 = read_U32
+
+let read_S32 input =
+  match Input.get0 input with
+  | None -> failwith "read_S32 error"
+  | Some c ->
+    let next_input = Input.sub_suffix 1 input in
+    let i = Char.code c in
+    let i = if i < 64 then i else i - 128 in
+    (* 64=2⁶ *)
+    (i, next_input)
+
+let read_S64 = read_S32
+
+let deserialize_indice = read_U32
+
+let deserialize_numtype input =
+  let b, next_input = read_byte input in
+  match b with
+  | '\x7F' -> (Some (Num_type I32), next_input)
+  | '\x7E' -> (Some (Num_type I64), next_input)
+  | '\x7D' -> (Some (Num_type F32), next_input)
+  | '\x7C' -> (Some (Num_type F64), next_input)
+  | _ -> (None, input)
+
+let deserialize_vectype input =
+  let b, _next_input = read_byte input in
+  match b with (* | '\x7B' -> (Some V128, input) *)
+  | _ -> (None, input)
+
+let deserialize_reftype input =
+  let b, _next_input = read_byte input in
+  match b with
+  (* | '\x70' -> (Some FUNCREF, input) *)
+  (* | '\x6F' -> (Some EXTERNREF, input) *)
+  | _ -> (None, input)
+
+let deserialize_valtype input =
+  match deserialize_numtype input with
+  | Some t, input -> (t, input)
+  | None, _ -> (
+    match deserialize_vectype input with
+    | Some t, input -> (t, input)
+    | None, _ -> (
+      match deserialize_reftype input with
+      | Some t, input -> (t, input)
+      | None, _ -> failwith "deserialize_valtype error" ) )
```

Comment:
It could be possible to get something like:


```ocaml
let deserialize_valtype input =
  match deserialize_numtype input with
  | Ok _ as v -> v
  | Error () ->
    match deserialize_vectype input with
    | Ok _ as v -> v
    | Error () ->
      match deserialize_reftype input with
      | Ok _ as v -> v
      | Error () -> Error "deserialize_valtype error"
```